Flux GA release spec and long term support pledge (#3254)
fluxcd/flux2: docs/release/flux.md

Release candidates are intended for testing new features or improvements before a final release.

In most cases, a maintainer will publish a release candidate for Flux users to test on their
staging clusters. Release candidates are not meant to be deployed in production unless advised
to do so by a maintainer.

Release candidates can be unstable and they are deprecated by subsequent RC or stable versions.

### Patch releases

Patch releases are intended for critical bug fixes to the latest minor version,
such as addressing security vulnerabilities or fixes to severe problems with no workaround.

Patch releases do not contain breaking changes, feature additions or any type of user-facing changes.
If a CVE fix requires a breaking change, then a minor release will provide the fix.

We expect users to be running the latest patch release of a given minor release.

### Minor releases

Minor releases are intended for backward-compatible feature additions and improvements.
Note that breaking changes may occur if required by a security vulnerability fix.

Minor releases are used when updating the Flux controllers or Kubernetes dependencies
from one minor version to another.

> In effect, this means a Flux minor version will be released at least every four months after each
> Kubernetes minor version release. To properly validate the Flux CLI and controllers against
> the latest Kubernetes version, we reserve a time window of at least two weeks for end-to-end testing. 

Reviewer: ```suggestion
In effect, this means a Flux minor version will be released at least every four months, after each
Kubernetes minor version release. For in-depth information about this, please refer to the
[release cadence](#release-cadence) section of this document.
```

### Major releases

Major releases are intended for drastic changes to the Flux behaviour or security stance.

A Flux major release will be announced ahead of time throughout all communication channels,
and a support window of one year will be provided for the previous major version.

## Release cadence

> Flux is at least released at the same rate as Kubernetes, following their cadence of three

Reviewer: ```suggestion
Flux is _at least_ released at the same rate as Kubernetes, following their cadence of three
```

same as above.

# Flux release spec

The Flux project repository [fluxcd/flux2](https://github.com/fluxcd/flux2) contains
the Flux command-line tool source code and the Kubernetes manifests for
bundling the [Flux controllers](controllers.md) into a distributable package.

## Release versioning

Flux is released by following the [semver](https://semver.org/) conventions:

> - `vX.Y.Z-RC.W` release candidates e.g. `v2.0.0-rc.1`

Reviewer 2: ```suggestion
- `vX.Y.Z-rc.W` release candidates e.g. `v2.0.0-rc.1`
```

# Flux release spec

The Flux project repository [fluxcd/flux2](https://github.com/fluxcd/flux2) contains
the Flux command-line tool source code and the Kubernetes manifests for
bundling the [Flux controllers](controllers.md) into a distributable package.

## Release versioning

Flux is released by following the [semver](https://semver.org/) conventions:

> - `vX.Y.Z-RC.W` release candidates e.g. `v2.0.0-RC.1`
> - `vX.Y.Z` stable releases e.g. `v2.0.0`
> 
> The Flux project maintains release branches for the most recent three minor releases
> e.g. `release-2.0`, `release-2.1` and `release-2.2`.

Reviewer: ```suggestion
- `vX.Y.Z-RC.W` release candidates e.g. `v2.0.0-rc.1`
- `vX.Y.Z` stable releases e.g. `v2.0.0`

The Flux project maintains release branches for the most recent three minor releases
e.g. `release/2.0.x`, `release/2.1.x` and `release/2.2.x`.
```

- We have tagged our RC with `-rc.X`, so lets use that as a proper example.
- I prefer `/` "namespacing" on repositories where features are also developed in branches, and `-` is used in those names. This also eases things like branch protection and automation without creating false positives.

### Release candidates

> Release candidates are intended for testing new features or improvements before a final release.
> 
> In most cases, a maintainer will publish a release candidate for Flux users to tests it on their
> staging clusters. Release candidates are not meant to be deployed in production unless advised
> to do so by a maintainer.
> 
> Release candidates can be unstable and they are deprecated by subsequent RC or stable version.

Reviewer 3: ```suggestion
Release candidates are intended for testing new features or improvements before a final release.

In most cases, a maintainer will publish a release candidate for Flux users to test on their
staging clusters. Release candidates are not meant to be deployed in production unless advised
to do so by a maintainer.

Release candidates can be unstable and they are deprecated by subsequent RC or stable versions.
```

### Patch releases

Patch releases are intended for critical bug fixes to the latest minor version,
such as addressing security vulnerabilities or fixes to severe problems with no workaround.

Patch releases do not contain breaking changes, feature additions or any type of user-facing changes.
If a CVE fix requires a breaking change, then a minor release will provide the fix.

We expect users to be running the latest patch release of a given minor release.

### Minor releases

> Minor releases are intended for backwards compatible feature additions and improvements.
> Note that breaking changes may occur if required by a security vulnerability fix.
> 
> Minor releases are used when updating the Flux controllers or Kubernetes dependencies
> from one minor version to another.
> 
> In effect, this means a Flux minor version will be released at least every four months after each
> Kubernetes minor version release. To properly validate the Flux CLI and controllers against
> the latest Kubernetes version, we reserve a time window of at least two weeks for end-to-end testing. 

Reviewer 3: ```suggestion
Minor releases are intended for backward-compatible feature additions and improvements.
Note that breaking changes may occur if required by a security vulnerability fix.

Minor releases are used when updating the Flux controllers or Kubernetes dependencies
from one minor version to another.

In effect, this means a Flux minor version will be released at least every four months after each
Kubernetes minor version release. To properly validate the Flux CLI and controllers against
the latest Kubernetes version, we reserve a time window of at least two weeks for end-to-end testing. 
```

### Major releases

Major releases are intended for drastic changes to the Flux behaviour or security stance.

A Flux major release will be announced ahead of time throughout all communication channels,
and a support window of one year will be provided for the previous major version.

## Release cadence

> Flux is at least released at the same rate as Kubernetes, following their cadence of three
> minor releases per year. After each Kubernetes minor release, the CLI and all controllers are
> tested against the latest Kubernetes version and are released at approximately two weeks after Kubernetes.
> The newly released Flux version offers support for Kubernetes N-2 minor versions.
> 
> Flux may have more than three minor releases per year, if maintainers decide to ship a 
> new feature or optimisation ahead of schedule.

Reviewer 3: ```suggestion
Flux is at least released at the same rate as Kubernetes, following their cadence of three
minor releases per year. After each Kubernetes minor release, the CLI and all controllers are
tested against the latest Kubernetes version and are released approximately two weeks after Kubernetes.
The newly released Flux version offers support for Kubernetes N-2 minor versions.

Flux may have more than three minor releases per year if maintainers decide to ship a 
new feature or optimization ahead of schedule.
```
